src/functions/ensureEnv.ts

```diff
@@ -0,0 +1,161 @@
+import * as clc from "cli-color";
+
+import { configstore } from "../configstore";
+import { FirebaseError } from "../error";
+import { promptOnce } from "../prompt";
+import { logBullet, logWarning } from "../utils";
+import * as backend from "../deploy/functions/backend";
+import * as deploymentTool from "../deploymentTool";
+import * as envstore from "./envstore";
+import * as getProjectId from "../getProjectId";
+import * as helper from "../deploy/functions/functionsDeployHelper";
+
+const ENVSTORE_INTERNAL_ID = "firebase-functions-internal";
+const CONFIGSTORE_KEY = "envstore";
+const CONFIGSTORE_TTL = 1000 * 60 * 60 * 24; // 1 day
+const DEFAULT_ENV_KEYS = ["FIREBASE_CONFIG"];
+
+/**
+ * Check if the EnvStore API is active.
+ *
+ * This is an bespoke method of checking whether user has opted in
+ * to activate the EnvStore API.
+ *
+ * We check for existance of non-emtpy ENV_ID=${ENVSTORE_INTERNAL_ID}
+ * which is created when user "enables" the EnvStore API.
+ */
+async function _check(projectId: string): Promise<boolean> {
+  const resp = await envstore.getStore(projectId, ENVSTORE_INTERNAL_ID);
+  return !!resp.vars;
+}
+
+/**
+ * Cached check for active EnvStore API.
+ */
+export async function check(projectId: string): Promise<boolean> {
+  // Check actice state from local cache.
+  const cached = configstore.get(CONFIGSTORE_KEY) as { lastActiveAt: string } | undefined;
+  if (cached?.lastActiveAt) {
+    const activeAt = new Date(cached.lastActiveAt);
+    const diff = Date.now() - activeAt.getTime();
+    if (diff <= CONFIGSTORE_TTL) {
+      return true;
+    }
+    configstore.delete(CONFIGSTORE_KEY);
+  }
+
+  // Query the EnvStore API to check active state.
+  const checked = await _check(projectId);
+  if (checked) {
+    configstore.set(CONFIGSTORE_KEY, { lastActiveAt: Date.now() });
+    return true;
+  }
+  return false;
+}
+
+/**
+ * Enable the EnvStore API.
+ *
+ * This is an bespoke method of "enabling" EnvStore API. We "enable" the
+ * EnvStore API setting up a non-empty EnvStore ENV_ID=${ENVSTORE_INTERNAL_ID}.
+ */
+export async function enable(projectId: string): Promise<void> {
+  await envstore.patchStore(projectId, ENVSTORE_INTERNAL_ID, { ENABLED: "1" });
+}
+
+interface UserEnv {
+  fnLabel: string;
+  envs: Record<string, string>;
+}
+
+/**
+ * Lookup existing cloud functions and collect user-defined env variables.
+ *
+ * If user's project hasn't activated the EnvStore API, we consider
+ * all non-default environment variable as "user-defined". The user-defined
+ * env is significant because they will be removed on the next deploy following
+ * EnvStore API activation.
+ */
+async function getUserEnvs(projectId: string): Promise<UserEnv[]> {
+  const have = await backend.existingBackend({ projectId, filters: [] }, /* forceRefresh= */ true);
```

Comment:
1. Will this be used during a CLI command that also has a context object? The context object is also used to store a cache.
2. If not, you don't need forceRefresh because this context object doesn't have the cache fields.

Reply:
(2) - changing the passed value to `false`.